### PR TITLE
Add Volume Snapshot Support

### PR DESCRIPTION
**Pull request by xing-yang**
This document proposes to add Kubernetes snapshot support in-tree.

### CODE REVIEW

File: contributors/design-proposals/storage/volume-snapshotting.md [hunk 1]
```diff
@@ -21,498 +25,529 @@ For the first version of snapshotting support in Kubernetes, only on-demand snap
...
+
+* Reconciliation loop in the controller
+
+    * For every `VolumeSnapshot` in the ActualStateOfWorld (ASW) find the corresponding `VolumeSnapshot` in the DSW.
+    If such a snapshot does not exist, start a snapshot deletion operation:
+
+        * Determine the correct volume snapshot plugin to use from the `VolumeSnapshotData` referenced by the
+        `VolumeSnapshot`
+
+        * Create a delete operation: only one such operation is allowed to exist for the given `VolumeSnapshot` and
```
Comment (mlmhl) [inline contributors/design-proposals/storage/volume-snapshotting.md]: If the controller stopped before we delete `VolumeSnapshotData`, according `VolumeSnapshot` no longer exists in `ASW` after controller started again, and the `VolumeSnapshotData` will be orphaned. Maybe we need an asynchronous go routine to clean up these `VolumeSnapshotDatas`.

**Comment by jingxu97** [comment]
cc @thockin PTAL

**Comment by xing-yang** [comment]
We are going to use CRD instead of in-tree APIs. A new PR is submitted here: https://github.com/kubernetes/community/pull/2335.  Closing this one.

**Pull request closed without merging**